### PR TITLE
Introduces more standard types wasm.Index, wasm.NameMap etc.

### PR DESCRIPTION
**Pull request by codefromthecrypt**
This reworks the name section to be closer to the binary format

This makes the name section data structure cheaper and more closely
matches the binary structure. Doing this allows us to keep the original
decode order and avoid expense hashing or re-ordering redundantly.

This also changes language to match the word local section in
preparation of module defined functions. It also introduces `wasm.Index`
to avoid repeating docs so often, Finally, this backfills some more
tests.

As we don't re-sort etc, benchmark performance has improved, particularly in the round-trip case

Before:
```
BenchmarkCodecExample/binary.DecodeModule
BenchmarkCodecExample/binary.DecodeModule-16         	  210984	      4853 ns/op	    2784 B/op	     168 allocs/op
BenchmarkCodecExample/binary.EncodeModule
BenchmarkCodecExample/binary.EncodeModule-16         	  240454	      4977 ns/op	    3832 B/op	     143 allocs/op
BenchmarkCodecExample/text.DecodeModule
BenchmarkCodecExample/text.DecodeModule-16           	  101834	     11375 ns/op	    6600 B/op	     248 allocs/op
BenchmarkCodecExample/wat2wasm_via_text.DecodeModule->binary.EncodeModule
BenchmarkCodecExample/wat2wasm_via_text.DecodeModule->binary.EncodeModule-16         	   67006	     16439 ns/op	   10416 B/op	     391 allocs/op
BenchmarkCodecExample/wat2wasm_via_wasmtime.Wat2Wasm
BenchmarkCodecExample/wat2wasm_via_wasmtime.Wat2Wasm-16                              	   43602	     27315 ns/op	    1152 B/op	       3 allocs/op
```

Now:
```
BenchmarkCodecExample/binary.DecodeModule
BenchmarkCodecExample/binary.DecodeModule-16         	  254241	      4706 ns/op	    2304 B/op	     179 allocs/op
BenchmarkCodecExample/binary.EncodeModule
BenchmarkCodecExample/binary.EncodeModule-16         	  325231	      3603 ns/op	    3408 B/op	     127 allocs/op
BenchmarkCodecExample/text.DecodeModule
BenchmarkCodecExample/text.DecodeModule-16           	  108086	     10907 ns/op	    5928 B/op	     259 allocs/op
BenchmarkCodecExample/wat2wasm_via_text.DecodeModule->binary.EncodeModule
BenchmarkCodecExample/wat2wasm_via_text.DecodeModule->binary.EncodeModule-16         	   80340	     14828 ns/op	    9336 B/op	     386 allocs/op
```

### CODE REVIEW

File: wasm/binary/names_test.go [hunk 1]
```diff
@@ -103,34 +109,6 @@ func TestEncodeNameSectionData(t *testing.T) {
 				0x01, 0x01, 'r', // index 1, size of "r", "r"
 			},
 		},
-		{
```
Comment (codefromthecrypt) [inline wasm/binary/names_test.go:106]: we decode in order and don't pass anything into a map that would lose the order, so we don't need this anymore

File: wasm/example/testdata/example.wat [hunk 2]
```diff
@@ -5,7 +5,7 @@
 ;; * start function
 (module $example
 	(type $i32i32_i32 (func (param i32 i32) (result i32)))
-	(import "wasi_snapshot_preview1" "arg_sizes_get" (func $runtime.arg_sizes_get (type $i32i32_i32)))
```
Comment (codefromthecrypt) [inline wasm/example/testdata/example.wat:8]: this was an old typo

File: wasm/store.go [hunk 3]
```diff
@@ -529,22 +529,37 @@ func (s *Store) buildFunctionInstances(module *Module, target *ModuleInstance) (
 	memoryDeclarations = append(memoryDeclarations, module.MemorySection...)
 	tableDeclarations = append(tableDeclarations, module.TableSection...)
 
-	var functionNames map[uint32]string
-	if module.NameSection != nil && module.NameSection.FunctionNames != nil {
+	// Convert the any NameSection.FunctionNames names to a map for easy lookup
```
Comment (codefromthecrypt) [inline wasm/store.go:532]:
```suggestion
```

File: wasm/store_test.go [hunk 4]
```diff
@@ -20,3 +20,35 @@ func TestGetModuleInstance(t *testing.T) {
 	m2 := s.getModuleInstance(name)
 	require.Equal(t, m1, m2)
 }
+
+func TestBuildFunctionInstances_FunctionNames(t *testing.T) {
```
Comment (codefromthecrypt) [inline wasm/store_test.go:24]: this was pretty fun not sarcastically

**Review by mathetake — APPROVED** [review]

**Comment by codefromthecrypt** [comment]
Thanks for lending an eye!

**Pull request merged**